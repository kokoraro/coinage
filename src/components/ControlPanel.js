import React, { Component } from 'react';
import AddCoinListItem from './AddCoinListItem';

class ControlPanel extends Component {
    constructor(props) {
        super(props);
        this.state = {
            selectCoins: null,
        };
        this.handleAddSubmit = this.handleAddSubmit.bind(this);
    }

    componentDidMount() {
        const { selectCoins } = this.props;
        this.setState({ selectCoins });
    }

    componentWillReceiveProps(nextProps) {
        const { selectCoins } = this.props;
        if (nextProps.selectCoins !== selectCoins) {
            this.setState({ selectCoins });
        }
    }

    handleAddCoinClick(id) {
        const { selectCoins } = this.state;
        const newSelectCoins = selectCoins.map(c => (
            c.id === id ? { ...c, showing: !c.showing } : c
        ));
        this.setState({ selectCoins: newSelectCoins });
    }

    handleAddSubmit() {
        const { selectCoins } = this.state;
        const { handleAddCoins } = this.props;
        const newSelectCoins = selectCoins
            .filter(c => c.showing)
            .map(c => c.id);
        handleAddCoins(newSelectCoins);
    }

    render() {
        const {
            addOpen,
            toggleAddOpen,
            handleCurrency,
            handleRefresh,
        } = this.props;
        const { selectCoins } = this.state;
        const form = addOpen ? (
            <div>
                <h4>choose coins to add...</h4>
                <button
                    type="button"
                    onClick={this.handleAddSubmit}
                    className="add-coins-submit"
                >
                    Submit
                </button>
                <div className="add-coins">
                    {selectCoins.map(c => (
                        <AddCoinListItem
                            handleAddCoinClick={() => this.handleAddCoinClick(c.id)}
                            key={c.id}
                            data={c}
                        />
                    ))}
                </div>
                <button
                    type="button"
                    onClick={this.handleAddSubmit}
                    className="add-coins-submit"
                >
                    Submit
                </button>
            </div>
        ) : null;

        return (
            <div>
                <div className="control-panel">
                    <button
                        type="button"
                        className="add-button"
                        onClick={toggleAddOpen}
                    >
                        Edit
                    </button>
                    <button type="button" onClick={handleCurrency}>Currency</button>
                    <button type="button" onClick={handleRefresh}>Refresh</button>
                </div>
                {form}
            </div>
        );
    }
}

export default ControlPanel;
