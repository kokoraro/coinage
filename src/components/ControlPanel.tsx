import React, { Component } from 'react';
import AddCoinListItem from './AddCoinListItem';
import { Coin } from '../redux/coins/types';

interface ControlPanelProps {
    selectCoins: Coin[];
    handleAddCoins: (coins: number[]) => void;
    addOpen: boolean;
    toggleAddOpen: () => void;
    handleCurrency: () => void;
    handleRefresh: () => void;
}

interface ControlPanelState {
    selectCoins: Coin[];
}

class ControlPanel extends Component<ControlPanelProps, ControlPanelState> {
    constructor(props: ControlPanelProps) {
        super(props);
        this.state = {
            selectCoins: [],
        };
        this.handleAddSubmit = this.handleAddSubmit.bind(this);
    }

    componentDidMount() {
        const { selectCoins } = this.props;
        this.setState({ selectCoins });
    }

    componentWillReceiveProps(nextProps: ControlPanelProps) {
        const { selectCoins } = this.props;
        if (nextProps.selectCoins !== selectCoins) {
            this.setState({ selectCoins });
        }
    }

    handleAddCoinClick(id: number) {
        const { selectCoins } = this.state;
        const newSelectCoins = selectCoins.map(c => (
            c.id === id ? { ...c, showing: !c.showing } : c
        ));
        this.setState({ selectCoins: newSelectCoins });
    }

    handleAddSubmit() {
        const { selectCoins } = this.state;
        const { handleAddCoins } = this.props;
        const newSelectCoins = selectCoins
            .filter(c => c.showing)
            .map(c => c.id);
        handleAddCoins(newSelectCoins);
    }

    render() {
        const {
            addOpen,
            toggleAddOpen,
            handleCurrency,
            handleRefresh,
        } = this.props;
        const { selectCoins } = this.state;

        return (
            <div>
                <div className="control-panel">
                    <button
                        type="button"
                        className="add-button"
                        onClick={toggleAddOpen}
                    >
                        Edit
                    </button>
                    <button type="button" onClick={handleCurrency}>Currency</button>
                    <button type="button" onClick={handleRefresh}>Refresh</button>
                </div>
                {addOpen && (
                    <div>
                        <h4>choose coins to add...</h4>
                        <button
                            type="button"
                            onClick={this.handleAddSubmit}
                            className="add-coins-submit"
                        >
                            Submit
                        </button>
                        <div className="add-coins">
                            {selectCoins.map(c => (
                                <AddCoinListItem
                                    handleAddCoinClick={() => this.handleAddCoinClick(c.id)}
                                    key={c.id}
                                    data={c}
                                />
                            ))}
                        </div>
                        <button
                            type="button"
                            onClick={this.handleAddSubmit}
                            className="add-coins-submit"
                        >
                            Submit
                        </button>
                    </div>
                )}
            </div>
        );
    }
}

export default ControlPanel;
