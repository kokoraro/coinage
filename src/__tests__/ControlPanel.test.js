import React from 'react';
import { shallow } from 'enzyme';
import ControlPanel from '../components/ControlPanel';

const controlPanel = shallow(<ControlPanel />);

describe("ControlPanel component", () => {
  it("renders correctly", () => {
    expect(controlPanel).toMatchSnapshot();
  });

  it("add click shows input", () => {
    expect(controlPanel.find('input').length).toBe(0);
    controlPanel.find('.add-button').simulate('click');
    expect(controlPanel.find('input').length).toBe(1);
  });

});