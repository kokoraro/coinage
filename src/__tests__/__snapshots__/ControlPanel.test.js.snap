// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlPanel component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ControlPanel />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="control-panel"
        >
          <button
            className="add-button"
            onClick={[Function]}
          >
            Add
          </button>
          <button
            onClick={undefined}
          >
            Currency
          </button>
          <button
            onClick={undefined}
          >
            Refresh
          </button>
        </div>,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="add-button"
              onClick={[Function]}
            >
              Add
            </button>,
            <button
              onClick={undefined}
            >
              Currency
            </button>,
            <button
              onClick={undefined}
            >
              Refresh
            </button>,
          ],
          "className": "control-panel",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add",
              "className": "add-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Currency",
              "onClick": undefined,
            },
            "ref": null,
            "rendered": "Currency",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Refresh",
              "onClick": undefined,
            },
            "ref": null,
            "rendered": "Refresh",
            "type": "button",
          },
        ],
        "type": "div",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="control-panel"
          >
            <button
              className="add-button"
              onClick={[Function]}
            >
              Add
            </button>
            <button
              onClick={undefined}
            >
              Currency
            </button>
            <button
              onClick={undefined}
            >
              Refresh
            </button>
          </div>,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="add-button"
                onClick={[Function]}
              >
                Add
              </button>,
              <button
                onClick={undefined}
              >
                Currency
              </button>,
              <button
                onClick={undefined}
              >
                Refresh
              </button>,
            ],
            "className": "control-panel",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add",
                "className": "add-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Currency",
                "onClick": undefined,
              },
              "ref": null,
              "rendered": "Currency",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Refresh",
                "onClick": undefined,
              },
              "ref": null,
              "rendered": "Refresh",
              "type": "button",
            },
          ],
          "type": "div",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
